src/injected/analyzers/tab-stops-orchestrator.ts

```diff
@@ -0,0 +1,111 @@
+import {
+    TabStopRequirementResult,
+    TabStopsRequirementEvaluator,
+} from 'injected/tab-stops-requirement-evaluator';
+import { TabbableElementGetter } from 'injected/tabbable-element-getter';
+import { FocusableElement } from 'tabbable';
+
+export class TabStopRequirementOrchestrator
+    implements AllFrameRunnerTarget<TabStopRequirementResult>
+{
+    public readonly name: string = 'TabStopRequirementOrchestrator';
+    private tabbableTabStops: FocusableElement[];
+    private actualTabStops: Set<HTMLElement> = new Set();
+    private latestVisitedTabStop: HTMLElement = null;
+    private reportResults: (payload: TabStopRequirementResult) => Promise<void>;
+
+    constructor(
+        private readonly dom: Document,
+        private readonly tabbableElementGetter: TabbableElementGetter,
+        private readonly windowUtils: WindowUtils,
+        private readonly tabStopsRequirementEvaluator: TabStopsRequirementEvaluator,
+        private readonly getUniqueSelector: (element: HTMLElement) => string,
+    ) {}
+
+    public start = () => {
+        this.dom.addEventListener('keydown', this.onKeydownForFocusTraps);
+        this.dom.addEventListener('focusin', this.addNewTabStop);
+        this.tabbableTabStops = this.tabbableElementGetter.getRawElements();
+        const tabbableFocusOrderResults =
+            this.tabStopsRequirementEvaluator.getTabbableFocusOrderResults(this.tabbableTabStops);
+        tabbableFocusOrderResults.forEach(result => {
+            this.reportResults(result);
+        });
+    };
+
+    public stop = () => {
+        this.dom.removeEventListener('keydown', this.onKeydownForFocusTraps);
+        this.dom.removeEventListener('focusin', this.addNewTabStop);
+        const keyboardNavigationResults =
+            this.tabStopsRequirementEvaluator.getKeyboardNavigationResults(
+                this.tabbableTabStops,
+                this.actualTabStops,
+            );
+        keyboardNavigationResults.forEach(this.reportResults);
+    };
+
+    public setResultCallback = (
+        reportResultCallback: (payload: TabStopRequirementResult) => Promise<void>,
+    ) => {
+        this.reportResults = reportResultCallback;
+    };
+
+    public transformChildResultForParent = (
+        result: TabStopRequirementResult,
+        messageSourceFrame: HTMLIFrameElement,
+    ): TabStopRequirementResult => {
+        const frameSelector = this.getUniqueSelector(messageSourceFrame);
+        result.selector = [frameSelector, ...result.selector];
+        return result;
+    };
+
+    private addNewTabStop = (focusEvent: FocusEvent) => {
+        const newTabStop = focusEvent.target as HTMLElement;
+
+        if (this.latestVisitedTabStop == null) {
+            this.actualTabStops.add(newTabStop);
+            this.latestVisitedTabStop = newTabStop;
+            return;
+        }
+
+        if (this.actualTabStops.has(newTabStop)) {
+            this.latestVisitedTabStop = newTabStop;
```

Comment:
Why don't we want to rungetFocusOrderResult when this tab element has been seen before? Are we just assuming that it will have been checked previously? Is it possible to get to the same newTabStop from two different lastVisitedTabStops?

Reply:
Good question. My understanding/thinking here is that there's a few reasons for this to happen but none would require us to get a 'focus-order' result. Effectively, when this happens we're seeing an element that we've seen before which could be three scenarios:

- we're stuck on this specific tab-stop => keyboard-trap; detected elsewhere.
- we're stuck in a loop including this tab-stop => keyboard-trap or keyboard-navigation issue, detected elsewhere.
- we've completed the full tab stop loop => we are done tabbing; detected elsewhere ... possibly.

Now, whether all these are behaving the way we expect is a different story but I think that's the intent.